src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingSubmissionService.java

```diff
@@ -194,31 +202,50 @@ private void updateGitDiffReport(ProgrammingExerciseParticipation programmingExe
-        if (isExamExercise && programmingExerciseParticipation instanceof ProgrammingExerciseStudentParticipation) {
-            var optionalStudent = ((ProgrammingExerciseStudentParticipation) programmingExerciseParticipation).getStudent();
-            Optional<User> optionalStudentWithGroups = optionalStudent.isPresent() ? userRepository.findOneWithGroupsAndAuthoritiesByLogin(optionalStudent.get().getLogin())
-                    : Optional.empty();
-            if (optionalStudentWithGroups.isPresent() && !examSubmissionService.isAllowedToSubmitDuringExam(programmingExercise, optionalStudentWithGroups.get(), true)) {
-                final String message = "The student " + optionalStudentWithGroups.get().getLogin()
-                        + " just illegally submitted code after the allowed individual due date (including the grace period) in the participation "
-                        + programmingExerciseParticipation.getId() + " for the exam programming exercise " + programmingExercise.getId();
+        // Students are not allowed to submit a programming exercise after the due date, if this happens we set the Submission to ILLEGAL
+        if (programmingExerciseParticipation instanceof ProgrammingExerciseStudentParticipation studentParticipation) {
+            var optionalStudent = studentParticipation.getStudent();
+            var optionalStudentWithGroups = optionalStudent.flatMap(student -> userRepository.findOneWithGroupsAndAuthoritiesByLogin(student.getLogin()));
+            if (optionalStudentWithGroups.isEmpty()) {
+                return;
+            }
+            User student = optionalStudentWithGroups.get();
+            if (!isAllowedToSubmit(studentParticipation, student, programmingSubmission)) {
+                final String message = "The student %s just illegally submitted code after the allowed individual due date (including the grace period) in the participation %d for the programming exercise %d"
+                        .formatted(student.getLogin(), programmingExerciseParticipation.getId(), programmingExercise.getId());
                 programmingSubmission.setType(SubmissionType.ILLEGAL);
                 programmingMessagingService.notifyInstructorGroupAboutIllegalSubmissionsForExercise(programmingExercise, message);
                 log.warn(message);
             }
         }
     }
 
+    private boolean isAllowedToSubmit(ProgrammingExerciseStudentParticipation participation, User studentWithGroups, ProgrammingSubmission programmingSubmission) {
+        ProgrammingExercise exercise = participation.getProgrammingExercise();
+        if (exercise.isExamExercise()) {
+            return examSubmissionService.isAllowedToSubmitDuringExam(exercise, studentWithGroups, true);
+        }
+        else {
+            return isAllowedToSubmitForCourseExercise(participation, programmingSubmission);
+        }
```

Comment:
```suggestion
        return isAllowedToSubmitForCourseExercise(participation, programmingSubmission);
```
to be consistent with the style used in `isAllowedToSubmitForCourseExercise` below